ouroboros-network/src/Ouroboros/Network/PeerSelection/LedgerPeers.hs

```diff
@@ -0,0 +1,188 @@
+  where
+    go :: StdGen -> Word16 -> [RelayAddress] -> m (StdGen, [RelayAddress])
+    go rng 0 picked = return (rng, picked)
+    go rng n picked =
+        let (r :: Word64, rng') = random rng
+            d = maxBound :: Word64
+            x = fromIntegral r % fromIntegral d
+            !pick_m = Map.lookupGE x pools in
+        case pick_m of
+             Nothing -> go rng' (n - 1) picked -- XXX We failed pick a peer. Shouldn't this be an error?
+             Just (ackStake, (stake, relays)) -> do
+                 let (ix, rng'') = randomR (0, NonEmpty.length relays - 1) rng'
+                     relay = relays NonEmpty.!! ix
+                 traceWith tracer $ PickedPeer relay ackStake stake
+                 go rng'' (n - 1) (relay : picked)
+
+-- | Example of how how peers from ledger could be used.
+-- In reality it will be implemented using the new connection-manager and governor.
+runLedgerPeers :: forall m.
+                      ( MonadAsync m
+                      , MonadDelay m
+                      , MonadMonotonicTime m
+                      )
+               => StdGen
+               -> Tracer m TraceLedgerPeers
+               -> LedgerPeersConsensusInterface m
+               -> (IPSubscriptionTarget -> m ())
+               -> (DnsSubscriptionTarget -> m ())
+               -> m ()
+runLedgerPeers inRng tracer LedgerPeersConsensusInterface{..} runIP runDns =
+    go inRng Nothing Map.empty
+  where
+
+    go rng oldTs_m peerMap = do
+        let peerListLifeTime = if Map.null peerMap then 30
+                                                   else 3600
+            useLedgerAfter = 4492800 + 2*21600 -- XXX two epochs after shelley hard fork
+        !now <- getMonotonicTime
+        (!peerMap', ts_m) <-
+            if isNothing oldTs_m || diffTime now (fromJust oldTs_m) > peerListLifeTime
+               then do
+                   pl_m <- atomically $ lpGetPeers useLedgerAfter
+                   !now' <- getMonotonicTime
+                   case pl_m of
+                        Nothing ->
+                            return (peerMap, Just now')
+                        Just pl -> do
+                            let pla = ackPoolStake pl
+                            traceWith tracer $ FetchingNewLedgerState $ Map.size pla
+                            return (pla, Just now')
+               else return (peerMap, oldTs_m)
+
+        (rng', !pickedPeers) <- pickPeers rng tracer peerMap' 8
+        traceWith tracer $ PickedPeers 8 pickedPeers
+        let (ipTarget, dnsTargets) = foldl' peersToSubTarget
+                                                 (IPSubscriptionTarget [] 0 , []) $ nub pickedPeers
+        ipAid <- async $ runIP ipTarget
+
+        dnsAids <- sequence [async $ runDns peer | peer <- dnsTargets]
+
+        -- let it run for 10 minutes, then repeat with a new set of random peers.
+        ttlAid <- async (threadDelay 600)
+
+        _ <- waitAnyCancel (ttlAid : ipAid : dnsAids)
```

Comment:
It would be nicer to use `race_`, or
```
withAsync (runIP ..) $ \ippAid ->
  withAsyncs ([runDns peer | peer <- dnsTargets]) $ \dnsAids ->
    race_ (threadDelay 600)
               (waitAny ...)
```
This is safe if an async exception hits the thread, or if any of the thread errors.
I think we have `withAsyncs` somewhere in the code.

Reply:
I agree that `withAsync` is the proper way to do this. But this code is unused and the connection-manager aware version will not  start any IP or DNS subscription works.

Reply:
I see your point, but I won't feel good merging this to master;  if I'd be a reviewer and stumbled upon such code, I would have more reservations for the rest of reviewing work.   There's an an easy fix too.

Reply:
The offending code has been removed.

Reply:
:100: